File: ruby/monosecret_rb/test/test_resolve.rb
# frozen_string_literal: true

require "json"
require "tmpdir"
require "minitest/autorun"

# Compile the native extension (statically linking libmonosecret_ffi.a) unless it
# is already built. ci-sdks.sh builds it explicitly; this covers standalone runs.
def ensure_ext
  pkg = File.expand_path("..", __dir__)
  return unless Dir[File.join(pkg, "lib", "monosecret", "monosecret_ext.{so,bundle}")].empty?

  system("bash", File.join(pkg, "scripts", "build-ext.sh")) || raise("build-ext.sh failed")
end

ensure_ext
require_relative "../lib/monosecret"

MANIFEST = <<~TOML
  [project]
  name = "rb-test"
  revision = "1.0"

  [profiles.default]
  DATABASE_URL = { description = "DB", required = true }
  DEV_SESSION_SECRET = { description = "Development-only session secret", required = false, default = "development-only-secret" }
  SENTRY_DSN = { description = "sentry", required = false }

  [scopes.database]
  secrets = ["DATABASE_URL"]
TOML

def project(dir, dotenv, manifest: MANIFEST)
  manifest_path = File.join(dir, "monosecret.toml")
  env_path = File.join(dir, ".env")
  File.write(manifest_path, manifest)
  File.write(env_path, dotenv)
  [manifest_path, "dotenv://#{env_path}"]
end

class ResolveTest < Minitest::Test
  def test_abi_version_nonempty
    refute_empty Monosecret.abi_version
  end

  def test_caller_context_is_structured_and_separate_from_reason
    builder = Monosecret::Monosecret.builder.with_caller(
      Monosecret::CallerContext.new(
        name: "git",
        version: "2.51.0",
        operation: "credential_get",
        resource: "github.com"
      )
    )
    request = builder.instance_variable_get(:@request)
    assert_equal "git", request.dig("caller", "name")
    assert_equal "credential_get", request.dig("caller", "operation")
    refute request.key?("reason")
  end

  def test_load_values_and_provenance
    Dir.mktmpdir do |dir|
      manifest, provider = project(dir, "DATABASE_URL=postgres://db\n")

      resolved = Monosecret.builder
                                       .with_path(manifest)
                                       .with_provider(provider)
                                       .with_reason("rb test")
                                       .load

      assert_equal "default", resolved.profile
      db = resolved.secrets["DATABASE_URL"]
      assert_equal "postgres://db", db.get
      assert_equal "provider", db.source
      refute_nil db.source_provider

      session = resolved.secrets["DEV_SESSION_SECRET"]
      assert_equal "development-only-secret", session.get
      assert_equal "default", session.source

      assert_equal ["SENTRY_DSN"], resolved.missing_optional
      refute resolved.secrets.key?("SENTRY_DSN")
    end
  end

  def test_inline_spec_resolves_at_its_logical_base_dir
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, "inline.env"), "TOKEN=inline-ruby\n")
      spec = {
        "project" => { "name" => "ruby-inline" },
        "providers" => { "env" => "dotenv://inline.env" },
        "profiles" => { "default" => { "secrets" => {
          "TOKEN" => { "description" => "token", "providers" => ["env"] }
        } } }
      }
      resolved = Monosecret::Monosecret.builder
                                        .with_inline_spec(spec, dir)
                                        .with_reason("ruby inline test")
                                        .load
      assert_equal "inline-ruby", resolved.secrets.fetch("TOKEN").get
    end
  end

  def test_set_as_env
    Dir.mktmpdir do |dir|
      manifest, provider = project(dir, "DATABASE_URL=postgres://db\n")
      ENV.delete("DATABASE_URL")

      Monosecret.builder
                            .with_path(manifest)
                            .with_provider(provider)
                            .with_reason("rb test")
                            .load
                            .set_as_env!

      assert_equal "postgres://db", ENV.fetch("DATABASE_URL")
    end
  end

  def test_scope_is_selected_and_returned
    Dir.mktmpdir do |dir|
      manifest, provider = project(
        dir,
        "DATABASE_URL=postgres://db\nSENTRY_DSN=https://sentry\n"
      )
      builder = Monosecret.builder
                                       .with_path(manifest)
                                       .with_provider(provider)
                                       .with_scope("database")
                                       .with_reason("rb scoped test")

      resolved = builder.load
      assert_equal "database", resolved.scope
      assert_equal ["DATABASE_URL"], resolved.secrets.keys

      report = builder.report
      assert_equal "database", report.scope
      assert_equal ["DATABASE_URL"], report.secrets.map(&:name)
    end
  end

  def test_missing_required_raises
    Dir.mktmpdir do |dir|
      manifest, provider = project(dir, "")

      error = assert_raises(Monosecret::MissingRequiredError) do
        Monosecret.builder
                              .with_path(manifest)
                              .with_provider(provider)
                              .with_reason("rb test")
                              .load
      end
      assert_includes error.missing, "DATABASE_URL"
    end
  end

  def test_as_path_returns_readable_file
    Dir.mktmpdir do |dir|
      manifest = <<~TOML
        [project]
        name = "rb-test"
        revision = "1.0"

        [profiles.default]
        TLS_CERT = { description = "cert", required = true, as_path = true }
      TOML
      manifest_path, provider = project(dir, "TLS_CERT=----cert----\n", manifest: manifest)

      # Block form closes the Resolved (removing the 0400 as_path temp file) so
      # the test leaves no secret-bearing file behind in the temp dir.
      Monosecret.builder
                            .with_path(manifest_path)
                            .with_provider(provider)
                            .with_reason("rb test")
                            .load do |resolved|
        cert = resolved.secrets["TLS_CERT"]
        assert cert.as_path
        assert_nil cert.value
        assert_equal "----cert----", File.read(cert.get)
      end
    end
  end

  def test_invalid_manifest_raises_error
    error = assert_raises(Monosecret::Error) do
      Monosecret.builder
                            .with_path("/definitely/does/not/exist/monosecret.toml")
                            .with_reason("rb test")
                            .load
    end
    refute_instance_of Monosecret::MissingRequiredError, error
    refute_empty error.kind
  end
end

# Cross-language conformance: resolve the shared fixtures and assert this SDK
# produces the canonical result every other SDK must also produce.
class ConformanceTest < Minitest::Test
  FIXTURES = File.expand_path("../../../conformance/fixtures", __dir__)
  CONSTRAINTS = File.expand_path("../../../conformance/constraint-violations", __dir__)

  def canonical(resolved)
    secrets = {}
    resolved.secrets.each do |name, secret|
      value = secret.as_path ? File.read(secret.get) : secret.value
      secrets[name] = { "value" => value, "source" => secret.source, "as_path" => secret.as_path }
    end
    {
      "profile" => resolved.profile,
      "secrets" => secrets,
      "missing_required" => [],
      "missing_optional" => resolved.missing_optional.sort
    }
  end

  def canonical_report(report)
    secrets = {}
    report.secrets.each do |s|
      secrets[s.name] = {
        "status" => s.status,
        "required" => s.required,
        "as_path" => s.as_path,
        "generated" => s.generated,
        "default_applied" => s.default_applied,
        # Present-or-not (not the path-dependent value) so the vector is
        # machine-independent yet still catches a dropped source_provider.
        "source_provider" => !s.source_provider.nil?
      }
    end
    { "profile" => report.profile, "secrets" => secrets }
  end

  def conformance_builder(dir)
    Monosecret.builder
                          .with_path(File.join(dir, "monosecret.toml"))
                          .with_provider("dotenv://#{File.join(dir, '.env')}")
                          .with_reason("conformance")
  end

  def test_conformance_typed_constraint_violations
    report = conformance_builder(CONSTRAINTS).report
    by_kind = report.constraint_violations.to_h { |violation| [violation.kind, violation] }

    assert_equal "cloud", by_kind.fetch("at_least_one").group
    assert_empty by_kind.fetch("at_least_one").present
    assert_equal "token", by_kind.fetch("exactly_one").group
    assert_equal %w[FALLBACK PRIMARY], by_kind.fetch("exactly_one").present
  end

  Dir.glob(File.join(FIXTURES, "*")).select { |p| File.directory?(p) }.sort.each do |dir|
    name = File.basename(dir)

    define_method("test_conformance_#{name}") do
      expected = JSON.parse(File.read(File.join(dir, "expected.json")))
      # Block form closes the Resolved so as_path temp files do not accumulate.
      conformance_builder(dir).load do |resolved|
        assert_equal expected, canonical(resolved)
      end
    end

    # Under no_values every SDK must emit the same all-null fields map: a
    # value-less secret serializes to null, not an empty string.
    define_method("test_conformance_no_values_#{name}") do
      expected = JSON.parse(File.read(File.join(dir, "expected_no_values.json")))
      conformance_builder(dir).with_no_values.load do |resolved|
        assert_equal expected, resolved.fields
      end
    end

    # The value-free report (status + provenance) is identical across SDKs.
    define_method("test_conformance_report_#{name}") do
      expected = JSON.parse(File.read(File.join(dir, "expected_report.json")))
      assert_equal expected, canonical_report(conformance_builder(dir).report)
    end
  end
end

File: python/monosecret_py/tests/test_resolve.py
"""Exercise the Python SDK end to end against the real C ABI."""

import pathlib

import pytest

from monosecret import (
    CallerContext,
    MissingRequiredError,
    Monosecret,
    MonosecretError,
    abi_version,
)

MANIFEST = """
[project]
name = "py-test"
revision = "1.0"

[profiles.default]
DATABASE_URL = { description = "DB", required = true }
DEV_SESSION_SECRET = { description = "Development-only session secret", required = false, default = "development-only-secret" }
SENTRY_DSN = { description = "sentry", required = false }

[scopes.database]
secrets = ["DATABASE_URL"]
"""


def _project(tmp_path: pathlib.Path, dotenv: str) -> tuple[str, str]:
    manifest_path = tmp_path / "monosecret.toml"
    env_path = tmp_path / ".env"
    manifest_path.write_text(MANIFEST)
    env_path.write_text(dotenv)
    return str(manifest_path), f"dotenv://{env_path}"


def test_abi_version_nonempty():
    assert abi_version()


def test_caller_context_is_structured_and_separate_from_reason():
    builder = Monosecret.builder().with_caller(
        CallerContext(
            name="git",
            version="2.51.0",
            operation="credential_get",
            resource="github.com",
        )
    )
    assert builder._request == {
        "caller": {
            "name": "git",
            "version": "2.51.0",
            "operation": "credential_get",
            "resource": "github.com",
        }
    }


def test_load_returns_values_and_provenance(tmp_path):
    manifest, provider = _project(tmp_path, "DATABASE_URL=postgres://db\n")

    resolved = (
        Monosecret.builder()
        .with_path(manifest)
        .with_provider(provider)
        .with_reason("py test")
        .load()
    )

    assert resolved.profile == "default"
    db = resolved.secrets["DATABASE_URL"]
    assert db.get == "postgres://db"
    assert db.source == "provider"
    assert db.source_provider is not None

    session = resolved.secrets["DEV_SESSION_SECRET"]
    assert session.get == "development-only-secret"
    assert session.source == "default"

    assert resolved.missing_optional == ["SENTRY_DSN"]
    assert "SENTRY_DSN" not in resolved.secrets


def test_inline_spec_resolves_at_its_logical_base_dir(tmp_path):
    env_path = tmp_path / "inline.env"
    env_path.write_text("TOKEN=inline-python\n")
    spec = {
        "project": {"name": "python-inline"},
        "providers": {"env": "dotenv://inline.env"},
        "profiles": {"default": {"secrets": {
            "TOKEN": {"description": "token", "providers": ["env"]},
        }}},
    }
    resolved = Monosecret.builder().with_inline_spec(spec, str(tmp_path)).with_reason(
        "python inline test"
    ).load()
    assert resolved.secrets["TOKEN"].get == "inline-python"


def test_scope_is_selected_and_returned(tmp_path):
    manifest, provider = _project(
        tmp_path,
        "DATABASE_URL=postgres://db\nSENTRY_DSN=https://sentry\n",
    )
    builder = (
        Monosecret.builder()
        .with_path(manifest)
        .with_provider(provider)
        .with_scope("database")
        .with_reason("py scoped test")
    )

    resolved = builder.load()
    assert resolved.scope == "database"
    assert list(resolved.secrets) == ["DATABASE_URL"]

    report = builder.report()
    assert report.scope == "database"
    assert [secret.name for secret in report.secrets] == ["DATABASE_URL"]


def test_set_as_env(tmp_path, monkeypatch):
    manifest, provider = _project(tmp_path, "DATABASE_URL=postgres://db\n")
    monkeypatch.delenv("DATABASE_URL", raising=False)

    resolved = (
        Monosecret.builder()
        .with_path(manifest)
        .with_provider(provider)
        .with_reason("py test")
        .load()
    )
    resolved.set_as_env()

    import os

    assert os.environ["DATABASE_URL"] == "postgres://db"


def test_missing_required_raises(tmp_path):
    manifest, provider = _project(tmp_path, "")  # DATABASE_URL absent

    with pytest.raises(MissingRequiredError) as exc:
        Monosecret.builder().with_path(manifest).with_provider(provider).with_reason(
            "py test"
        ).load()

    assert "DATABASE_URL" in exc.value.missing


def test_as_path_returns_readable_file(tmp_path):
    manifest_path = tmp_path / "monosecret.toml"
    env_path = tmp_path / ".env"
    manifest_path.write_text(
        """
[project]
name = "py-test"
revision = "1.0"

[profiles.default]
TLS_CERT = { description = "cert", required = true, as_path = true }
"""
    )
    env_path.write_text("TLS_CERT=----cert-bytes----\n")

    resolved = (
        Monosecret.builder()
        .with_path(str(manifest_path))
        .with_provider(f"dotenv://{env_path}")
        .with_reason("py test")
        .load()
    )

    try:
        cert = resolved.secrets["TLS_CERT"]
        assert cert.as_path
        assert cert.value is None
        assert pathlib.Path(cert.get).read_text() == "----cert-bytes----"
    finally:
        # as_path materializes a 0400 temp file the caller owns; remove it so
        # the test leaves no secret-bearing file behind in the temp dir.
        resolved.close()


def test_invalid_manifest_raises_monosecret_error(tmp_path):
    with pytest.raises(MonosecretError) as exc:
        Monosecret.builder().with_path(
            "/definitely/does/not/exist/monosecret.toml"
        ).with_reason("py test").load()

    assert not isinstance(exc.value, MissingRequiredError)
    assert exc.value.kind
